distutils/sysconfig.py

```diff
@@ -265,7 +268,7 @@ def customize_compiler(compiler):  # noqa: C901
     Mainly needed on Unix, so we can plug in the information that
     varies across Unices and is stored in Python's Makefile.
     """
-    if compiler.compiler_type == "unix":
+    if compiler.compiler_type in ["unix", "cygwin", "mingw32"]:
```

Comment:
I'd be okay with arbitrary changes to cygwinccompiler, but changes to sysconfig need a lot more scrutiny. This module is essentially deprecated and callers should prefer the stdlib `sysconfig` wherever possible. At the very least, it'd be great if there were some way to capture the expectation that's missed without these changes. Are there tests that could be run that fail on cygwin/mingw32 without these changes?

Reply:
> At the very least, it'd be great if there were some way to capture the expectation that's missed without these changes.

When building extensions the compiler flags used while building python itself should be used, similar to how it's on UNIX platforms, and without these changes, those are missing.

> Are there tests that could be run that fail on cygwin/mingw32 without these changes?

Yeah, the compiler flags change and that could be tested. For example, with this change on a MINGW64 env I get the following output
```sh
running build_ext
building 'cwrapper' extension
creating build
creating build/temp.mingw_x86_64-cpython-310
gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O2 -pipe -O3 -march=x86-64 -mtune=generic -O2 -pipe -O3 -IC:/msys64/mingw64/include/python3.10 -c cwrapper.c -o build/temp.mingw_x86_64-cpython-310/cwrapper.o
writing build/temp.mingw_x86_64-cpython-310/cwrapper.cp310-mingw_x86_64.def
gcc -shared -Wl,--enable-auto-image-base -pipe -pipe -s build/temp.mingw_x86_64-cpython-310/cwrapper.o build/temp.mingw_x86_64-cpython-310/cwrapper.cp310-mingw_x86_64.def -LC:/msys64/mingw64/lib/python3.10/config-3.10 -LC:/msys64/mingw64/lib -lpython3.10 -o C:/dev/mingw-distutils/cext/cwrapper.cp310-mingw_x86_64.pyd
```
while without these changes I get
```sh
running build_ext
building 'cwrapper' extension
creating build
creating build/temp.mingw_x86_64-cpython-310
gcc -mdll -O -Wall -IC:/msys64/mingw64/include/python3.10 -c cwrapper.c -o build/temp.mingw_x86_64-cpython-310/cwrapper.o
writing build/temp.mingw_x86_64-cpython-310/cwrapper.cp310-mingw_x86_64.def
gcc -shared -s build/temp.mingw_x86_64-cpython-310/cwrapper.o build/temp.mingw_x86_64-cpython-310/cwrapper.cp310-mingw_x86_64.def -LC:/msys64/mingw64/lib/python3.10/config-3.10 -LC:/msys64/mingw64/lib -lpython3.10 -o C:/dev/mingw-distutils/cext/cwrapper.cp310-mingw_x86_64.pyd
```

> Maybe now that this value is selecting on multiple, it should be isinstance(compiler, UnixCCompiler).

👍

Reply:
I meant to ask for tests that could be added to the test suite that capture this (missed/fixed) expectation.

Reply:
https://github.com/pypa/distutils/pull/184/commits/800fae90cb8171304c0a59f2157ebd52b005845d. There already exist tests to test the behaviour of this function and I've made it run on MinGW too.